Cannot find module '/my/path/to/package.json'

After running `pwmetrics` from path where `package.json` is absent we get exception kinda `Cannot find module '/my/path/to/package.json'`
Thanks to @pedro93 for finding this one 👍

--- a/bin/cli.js
+++ b/bin/cli.js
@@ -48,8 +48,7 @@ const cliFlags = yargs
   })
   .option('config', {
     'describe': 'Path to config file',
-    'type': 'string',
-    'default': 'package.json'
+    'type': 'string'
   })
   .option('upload', {
     'describe': 'Upload trace to Google Drive',
@@ -66,9 +65,13 @@ const cliFlags = yargs
     'type': 'boolean',
     'default': false
   })
-  .check( (argv) => {
-    // Make sure pwmetrics has been passed a url, either from cli or config file
-    if(argv.config) config = getConfigFromFile(argv.config);
+  .check((argv) => {
+    // Make sure pwmetrics has been passed a url, either from cli or config fileg()
+
+    // Test if flag was explicitly set, yargs default will always assume flag is called, lack of optional support
+    if (argv.config !== undefined) {
+      config = argv.config.length ? getConfigFromFile(argv.config) : getConfigFromFile();
+    }
 
     if (argv._.length === 0 && (config === undefined || !config.url))
       throw new Error(getMessageWithPrefix('ERROR', 'NO_URL'));

--- a/lib/utils/fs.ts
+++ b/lib/utils/fs.ts
@@ -4,6 +4,25 @@
 import * as path from 'path';
 
 function getConfigFromFile(fileName: string) {
+  if (fileName.endsWith('package.json')) return getConfigFromPkgFile(fileName);
+  else return getConfigFromFilePath(fileName);
+}
+
+const getConfigFromPkgFile = function(fileName: string) {
+  let resolved: string;
+  try {
+    const cwdPath = path.resolve(process.cwd(), fileName);
+    resolved = require.resolve(cwdPath);
+  } catch (e) {
+    // if there no package.json for current process.cwd then return empty config
+    return {};
+  }
+
+  const config = require(resolved);
+  return config.pwmetrics || {};
+};
+
+const getConfigFromFilePath = function(fileName: string) {
   let resolved: string;
   try {
     resolved = require.resolve(`./${fileName}`);
@@ -11,13 +30,9 @@ function getConfigFromFile(fileName: string) {
     const cwdPath = path.resolve(process.cwd(), fileName);
     resolved = require.resolve(cwdPath);
   }
-  const config = require(resolved);
-  if(config !== null && typeof config === 'object') {
-    if (resolved.endsWith('package.json'))
-      return config.pwmetrics || {};
-    else return config;
-  } else throw new Error(`Invalid config from ${fileName}`);
 
-}
+  const config = require(resolved);
+  return (config !== null && typeof config === 'object') ? config : {};
+};
 
 module.exports = { getConfigFromFile };

--- a/lib/utils/fs.ts
+++ b/lib/utils/fs.ts
@@ -3,26 +3,7 @@
 
 import * as path from 'path';
 
-function getConfigFromFile(fileName: string) {
-  if (fileName.endsWith('package.json')) return getConfigFromPkgFile(fileName);
-  else return getConfigFromFilePath(fileName);
-}
-
-const getConfigFromPkgFile = function(fileName: string) {
-  let resolved: string;
-  try {
-    const cwdPath = path.resolve(process.cwd(), fileName);
-    resolved = require.resolve(cwdPath);
-  } catch (e) {
-    // if there no package.json for current process.cwd then return empty config
-    return {};
-  }
-
-  const config = require(resolved);
-  return config.pwmetrics || {};
-};
-
-const getConfigFromFilePath = function(fileName: string) {
+function getConfigFromFile(fileName: string = 'package.json') {
   let resolved: string;
   try {
     resolved = require.resolve(`./${fileName}`);
@@ -30,9 +11,13 @@ const getConfigFromFilePath = function(fileName: string) {
     const cwdPath = path.resolve(process.cwd(), fileName);
     resolved = require.resolve(cwdPath);
   }
-
   const config = require(resolved);
-  return (config !== null && typeof config === 'object') ? config : {};
-};
+  if(config !== null && typeof config === 'object') {
+    if (resolved.endsWith('package.json'))
+      return config.pwmetrics || {};
+    else return config;
+  } else throw new Error(`Invalid config from ${fileName}`);
+  
+}
 
 module.exports = { getConfigFromFile };

--- a/readme.md
+++ b/readme.md
@@ -9,6 +9,7 @@
 ![image](https://cloud.githubusercontent.com/assets/39191/19417867/7aead922-93af-11e6-88ec-917dad6e89d2.png)
 
  Documentation on these metrics in the works. If you hit bugs in the metrics collection, report at [Lighthouse issues](https://github.com/GoogleChrome/lighthouse/issues).
+ [How to use article](https://medium.com/@denar90/easy-progressive-web-metrics-9afa5ed857c2)
 
 ### Install [![NPM pwmetrics package](https://img.shields.io/npm/v/pwmetrics.svg)](https://npmjs.org/package/pwmetrics)
 ```sh
@@ -304,7 +305,7 @@ module.exports = {
 ```js
 const PWMetrics = require('pwmetrics');
 
-const pwMetrics = new PWMetrics('http://example.com/', options); // [All available configuration options](#all-available-configuration-options) can be used as `opts`
+const pwMetrics = new PWMetrics('http://example.com/', options); // _All available configuration options_ can be used as `options`
 pwMetrics.start(); // returns Promise
 
 ```

--- a/test/cli.js
+++ b/test/cli.js
@@ -10,7 +10,7 @@ describe('CLI', function() {
   describe('url', () => {
     it('should throw error if a url is not provided by cli', () => {
       try {
-        childProcess.execSync('bin/cli.js --config');
+        childProcess.execSync('bin/cli.js');
       } catch (e) {
         expect(e.message).to.contain('No url entered..');
       }

--- a/test/cli.js
+++ b/test/cli.js
@@ -9,19 +9,15 @@ const expect = require('chai').expect;
 describe('CLI', function() {
   describe('url', () => {
     it('should throw error if a url is not provided by cli', () => {
-      try {
-        childProcess.execSync('bin/cli.js');
-      } catch (e) {
-        expect(e.message).to.contain('No url entered..');
-      }
+      return childProcess.exec('bin/cli.js', error => {
+        expect(error.message).to.contain('No url entered..');
+      });
     });
 
     it('should throw error if a url is not provided either by config or by cli', () => {
-      try {
-        childProcess.execSync('bin/cli.js --config=./test/fixtures/empty-config.js');
-      } catch (e) {
-        expect(e.message).to.contain('No url entered..');
-      }
+      return childProcess.exec('bin/cli.js --config=./test/fixtures/empty-config.js', error => {
+        expect(error.message).to.contain('No url entered..');
+      });
     });
 
     it('should not throw any error if a url is provided from cli', () => {

--- a/test/cli.js
+++ b/test/cli.js
@@ -2,6 +2,7 @@
 // Licensed under the Apache License, Version 2.0. See LICENSE
 'use strict';
 
+const path = require('path');
 const childProcess = require('child_process');
 const expect = require('chai').expect;
 
@@ -9,17 +10,32 @@ describe('CLI', function() {
   describe('url', () => {
     it('should throw error if a url is not provided by cli', () => {
       try {
-        childProcess.execSync('node bin/cli.js --config');
+        childProcess.execSync('bin/cli.js --config');
       } catch (e) {
         expect(e.message).to.contain('No url entered..');
       }
     });
+
     it('should throw error if a url is not provided either by config or by cli', () => {
       try {
-        childProcess.execSync('node bin/cli.js --config=./test/fixtures/empty-config.js');
+        childProcess.execSync('bin/cli.js --config=./test/fixtures/empty-config.js');
       } catch (e) {
         expect(e.message).to.contain('No url entered..');
       }
+    });
+
+    it('should not throw any error if a url is provided from cli', () => {
+      return childProcess.exec('bin/cli.js https://example.com', error => {
+        expect(error).to.be.null;
+      });
+    });
+
+    it('should not throw any error if a url is provided from cli where package.json is absent', () => {
+      return childProcess.exec('./cli.js https://example.com', {
+        cwd: path.join(process.cwd(), 'bin')
+      }, error => {
+        expect(error).to.be.null;
+      });
     });
   });
 });

--- a/test/fs.js
+++ b/test/fs.js
@@ -8,15 +8,15 @@ const expectations = require('./fixtures/run-expectations');
 
 describe('Config', () => {
   describe('validate input', () => {
-    context('from package.json', () => {
+    context('from package.json as default parameter', () => {
       it('should return empty object if pwmetrics property not defined', () => {
-        expect(getConfigFromFile('package.json')).to.be.deep.equal({});
+        expect(getConfigFromFile()).to.be.deep.equal({});
       });
     });
     context('with pwmetrics property defined', () => {
       it('should return the pwmetrics data', () => {
         packageJSON.pwmetrics = expectations;
-        expect(getConfigFromFile('package.json')).to.be.deep.equal(expectations);
+        expect(getConfigFromFile()).to.be.deep.equal(expectations);
         delete packageJSON.pwmetrics;
       });
     });
@@ -30,8 +30,8 @@ describe('Config', () => {
         expect(() => getConfigFromFile('./some/invalid/path')).to.throw(Error, 'Cannot find module');
       });
 
-      it('should return empty object if file in path is not a json', () => {
-        expect(getConfigFromFile('./test/fixtures/invalid-config')).to.be.deep.equal({});
+      it('should throw if file in path is not a json', () => {
+        expect(() => getConfigFromFile('./test/fixtures/invalid-config')).to.throw(Error, 'Invalid config from');
       });
     });
   });

--- a/test/fs.js
+++ b/test/fs.js
@@ -14,15 +14,9 @@ describe('Config', () => {
       });
     });
     context('with pwmetrics property defined', () => {
-      before(() => {
-        packageJSON.pwmetrics = expectations;
-      });
-
       it('should return the pwmetrics data', () => {
+        packageJSON.pwmetrics = expectations;
         expect(getConfigFromFile('package.json')).to.be.deep.equal(expectations);
-      });
-
-      after(() => {
         delete packageJSON.pwmetrics;
       });
     });
@@ -37,7 +31,7 @@ describe('Config', () => {
       });
 
       it('should return empty object if file in path is not a json', () => {
-        expect(() => getConfigFromFile('./test/fixtures/invalid-config')).to.throw(Error, 'Invalid config from');
+        expect(getConfigFromFile('./test/fixtures/invalid-config')).to.be.deep.equal({});
       });
     });
   });

